stdlib/SparseArrays/test/higherorderfns.jl

```diff
@@ -632,4 +632,11 @@ end
     @test minimum(sparse([1, 2], [1, 2], ones(Int32, 2)), dims = 1) isa Matrix
 end
 
+@testset "Issue #30118" begin
+    @test ((_, x) -> x).(Int, spzeros(3)) == spzeros(3)
+    @test ((_, _, x) -> x).(Int, Int, spzeros(3)) == spzeros(3)
+    @test ((_, _, _, x) -> x).(Int, Int, Int, spzeros(3)) == spzeros(3)
+    @test_broken ((_, _, _, _, x) -> x).(Int, Int, Int, Int, spzeros(3)) == spzeros(3)
```

Comment:
It looks like `capturescalars` has type inference issue when it has to recurse a lot (> 3).  I wounder if `capturescalars` could be implemented more naturally (for humans and maybe for Julia as well?) if you used `@generated` function.  Any reason why `capturescalars` is implemented based on recursion?